### ISSUE TITLE
Overhaul `sf::IpAddress`

### ISSUE THREAD

**Issue by vittorioromeo** [SFML/SFML#2138]
So apparently these tests pass:

```cpp
const sf::IpAddress ipAddress;
CHECK(ipAddress.toString() == "0.0.0.0"s);
CHECK(ipAddress.toInteger() == 0);
CHECK(ipAddress != sf::IpAddress::Any);

const sf::IpAddress any("0.0.0.0"s);
CHECK(any.toString() == "0.0.0.0"s);
CHECK(any.toInteger() == 0x00000000);
CHECK(any == sf::IpAddress::Any);
```

This is extremely worrisome to me -- because there's a hidden "invalid" state that cannot be distinguished from `toString` and `toInteger`.

So here's my proposal:

1. Rename `sf::IpAddress` to `sf::IpV4Address`. This can be done as a pure renaming refactoring PR.
2. `sf::IpV4Address` is an always-valid value semantic class. It's just a wrapper around a single `sf::Uint32`.
3. To create `sf::IpV4Address` instances, a factory function returning `std::optional<sf::IpV4Address>` must be used. The constructor of `sf::IpV4Address` is private.

Thoughts? @ChrisThrasher @eXpl0it3r @Bromeon

**Comment by ChrisThrasher** [comment]
I'd most prefer we just throw during construction. This is impossible for users to ignore, trivial to test, and maintains the clean constructor syntax that C++ developers expect. By returning an optional we're just imposing the error handling on the user who still has to correctly handle that optional or risk throwing a `std::bad_optional_access`. At least if we throw our own exception we can pick the error message such that it's as clear as possible why construction failed. This also means we get to maintain the other constructors like `sf::IpAddress(sf::Uint32)` which cannot fail without creating a weird split between fallible construction methods and infallible construction methods. 

I know this breaks with SFML convention but maybe we reconsider this design decision because as illustrated here, it's causing us to write messier interfaces. 

Exceptions aside, my 2nd favorite option would be `std::optional<sf::IpAddress> sf::IpAddress::create(/* all constructor overloads here */)` and remove all existing constructors, even those that cannot fail, for the sake of consistency. 

---

Unsure why we want to specify that these are IPv4 addresses unless we plan to add IPv6 addresses in the future. I suppose it's future proofing in case `sf::IpV6Address` gets added in a later 3.x minor release.

---

> To prevent a physical dependency between sf::IpV4Address and <string>/<string_view>/<optional>,...

I disagree with the idea that we need to optimize compile times to this degree. I think it's largely a waste of our time and makes SFML more difficult to maintain for very little benefit to users. If SFML were notorious for lengthy compile times I'd be more agreeable but on my 2021 laptop, a clean build of `master` including tests and examples takes 10 seconds with incremental builds being too fast to notice much of the time. 

I also dislike how, for example, you cannot use all of `<SMFL/Graphics/RectangleShape.hpp>`'s interface without also including `<SFML/Graphics/Texture.hpp>`. It's expected that if you include a public header, you don't have to include additional headers to use it yet we break that expectation in the name of faster compile times.

**Comment by Bromeon** [comment]
As mentioned before, I think the topic of error handling should be discussed for SFML 3 in general, as `sf::IpAddress` is not an isolated instance where it becomes relevant.

I have opened issue https://github.com/SFML/SFML/issues/2139 for that.

**Comment by Bromeon** [comment]
> Exceptions aside, my 2nd favorite option would be `std::optional<sf::IpAddress> sf::IpAddress::create(/* all constructor overloads here */)` and remove all existing constructors, even those that cannot fail, for the sake of consistency.

For completeness, using suggestion #5 from the error handling issue would be:
```cpp
#include <SFML/System/Result.hpp>

class IpAddress {
public:
    // "Primary" or "canonical" constructor
    IpAddress(sf::Uint8 a, sf::Uint8 b, sf::Uint8 c, sf::Uint8 d);

    // Named constructor idiom for other ways of instantiation, fallible or not
    static IpAddress fromInteger(sf::Uint32 ip);
    static Result<IpAddress> fromString(std::string_view str); // returns address or error (message/enum/...)
}
```

---



> I disagree with the idea that we need to optimize compile times to this degree. I think it's largely a waste of our time and makes SFML more difficult to maintain for very little benefit to users. If SFML were notorious for lengthy compile times I'd be more agreeable but on my 2021 laptop, a clean build of `master` including tests and examples takes 10 seconds with incremental builds being too fast to notice much of the time.

Even though I'm a fan of short compile times, I agree here. Since C++ doesn't have extension methods, we need to artificially split related functionality like `IpAddress` and `IpAddressUtils`. A user who wants to create an IP address cannot just look at the `sf::IpAddress` documentation, but also needs to look at `sf::IpAddressUtils`. This compile-time optimization becomes a leaky abstraction and has negative impacts on ergonomics.

**Comment by vittorioromeo** [comment]
> Even though I'm a fan of short compile times, I agree here. Since C++ doesn't have extension methods, we need to artificially split related functionality like IpAddress and IpAddressUtils. [...] This compile-time optimization becomes a leaky abstraction and has negative impacts on ergonomics.

I just want to make it clear that this subdivision is not merely a compile-time optimization, IMHO it is good engineering practice and an application of the SRP. An "ip address" class should not be concerned on conversions from a string -- that's a completely orthogonal responsibility that should go somewhere else. 

It's like having a `sf::Sprite` constructor taking a `sf::Image`. Convenient? Possibly. Good engineering? Not quite.

I would advise on performing this separation even if compilation times were not improved.

> A user who wants to create an IP address cannot just look at the sf::IpAddress documentation, but also needs to look at sf::IpAddressUtils. 

That is not entirely true. The `sf::IpAddress` documentation can freely mention and provide a hyperlink to `sf::IpAddressUtils`.

**Comment by binary1248** [comment]
From an IETF perspective there is no such thing as an "invalid IP address". All 2^32 distinct values for v4 and all 2^128 distinct values for v6 always represent an address. Whether an address makes sense in a given context depends on that specific context. I wouldn't needlessly complicate a class that is just meant to be a thin wrapper around the underlying 32-bit value.

The only reason why `m_valid` was necessary was because `sf::IpAddress` existed during the C++98 era when `std::optional` wasn't around yet. Certain address retrieval functions needed a way to indicate that no address was available to be returned. Today it is obvious that `std::optional` would be used for these cases, but back in the day the optional behaviour was combined into `sf::IpAddress` itself.

Since we have `std::optional` now I'd say just clean up the places where `sf::IpAddress::None` are used or address resolution can fail and just declare that every constructed `sf::IpAddress` is always valid. `sf::IpAddress::resolve()` should really be a static function that returns an optional.

- Remove default and string constructors
- Make resolve() a public static function that returns optional
- Return nullopt whenever an IP address is unavailable

**Comment by ChrisThrasher** [comment]
Closing because this is almost entirely addressed by #2145
